tools/provision

```diff
@@ -1,48 +1,57 @@
-    virtualenv $PYCUSTOMPATH_ARG "$BASEDIR/$VENVDIR"
-    echo "Virtualenv created."
-fi
-
-if [[ ! -d "$BASEDIR/$VENVDIR/$VENVEXECDIR" ]]; then
-    # POSIX compatibility layer and Linux environment emulation for Windows
-    # Virtual uses /Scripts instead of /bin on Windows.
-    # Read https://virtualenv.pypa.io/en/stable/userguide/
-    VENVEXECDIR="Scripts"
-fi
-
-source "$BASEDIR/$VENVDIR/$VENVEXECDIR/activate"
-RVAL=$?
-if [ "$RVAL" -ne 0 ]; then
-    echo "Failed to activate virtualenv."
-    exit
-fi
-
-# Install python dependencies if needed.
-cmp "$BASEDIR/$VENVDIR/installed-requirements.txt" requirements.txt 2>/dev/null
-RVAL=$? # Return value of the comparision. 0 means files are same.
-if [ "$RVAL" -ne 0 ]; then
-    pip install -r "$BASEDIR/requirements.txt"
-    cp -a requirements.txt "$BASEDIR/$VENVDIR/installed-requirements.txt"
-    echo "Requirements installed."
-fi
-echo 'Success!  Run `source' "$BASEDIR/$VENVDIR/$VENVEXECDIR/activate"'`' "to activate virtualenv."
+#!/usr/bin/env python
+
+import os
+import sys
+import pip
+import subprocess
+
+def main():
+    if len(sys.argv) > 1:
+        print("USAGE: {exec_path}\n"
+              "Creates a Python virtualenv. Its Python version is equal to "
+              "the Python version this command is executed with.".format(exec_path=__file__))
+        sys.exit(2)
+
+    base_dir = os.path.abspath(os.path.join(__file__, '..', '..'))
+    python_version = '.'.join(str(num) for num in sys.version_info[0:3])
+    python_interpreter = os.path.abspath(sys.executable)
+    venv_name = 'zulip-api-py{}-venv'.format(python_version)
+
+    venv_dir = os.path.join(base_dir, venv_name)
+    if not os.path.isdir(venv_dir):
+        if subprocess.call(['virtualenv', '-p', python_interpreter, venv_dir]):
+            raise OSError("The command `virtualenv -p {} {}` failed. Virtualenv not created!"
+                          .format(python_interpreter, venv_dir))
+        else:
+            print("New virtualenv created.")
+    else:
+        print("Virtualenv already exists.")
+
+    if os.path.isdir(os.path.join(venv_dir, 'Scripts')):
+        # POSIX compatibility layer and Linux environment emulation for Windows
+        # Virtual uses /Scripts instead of /bin on Windows.
+        # Read https://virtualenv.pypa.io/en/stable/userguide/
+        venv_exec_dir = 'Scripts'
+    else:
+        venv_exec_dir = 'bin'
+
+    # In order to install all required packages for the venv, we need to activate it. Since
+    # the activation script sets environmental variables, it needs to be executed inline with
+    # `exec`. The arguments `globals(), locals()` are required for Python 2.7 compatibility.
+    # `exec` cannot execute shell scripts, so we need to call `activate_this.py` instead of `activate`.
+    activate_path = os.path.abspath(os.path.join(venv_dir, venv_exec_dir, 'activate_this.py'))
+    exec(compile(open(activate_path).read(), activate_path, 'exec'), globals(), locals())
```

Comment:
Oh, I haven't seen that code piece. Can't tell which one is better off of my head, but I got the `exec(compile` version from a Python2/3 compatibility site, and tested with both python versions.